js/mage/adminhtml/grid-double-scroll.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * OpenMage
+ *
+ * This source file is subject to the Academic Free License (AFL 3.0)
+ * that is bundled with this package in the file LICENSE_AFL.txt.
+ * It is also available at https://opensource.org/license/afl-3-0-php
+ *
+ * @category    Mage
+ * @package     Mage_Adminhtml
+ * @copyright   Copyright (c) 2019-2023 The OpenMage Contributors (https://www.openmage.org)
+ * @license     https://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
+ */
+
+class gridDoubleScroll {
+    oldCallbacks = {};
+    _scrolling = false;
+
+    constructor(containerId, grid) {
+        this.containerId = containerId;
+        this.grid = grid;
+        this.setOldCallback('init', grid.initCallback);
+        this.grid.initCallback = this.onGridInit.bind(this);
+
+        this.handlerSyncWrapperScrollBar = this.syncWrapperScrollBar.bind(this);
+        this.handlerSyncScrollBarTop = this.syncScrollBarTop.bind(this);
+        this.handlerUpdateDoubleScrollWidth = this.updateDoubleScrollWidth.bind(this);
+        
+        this.initDoubleScroll();
+    }
+
+    initDoubleScroll() {
+        this.wrapperScrollBar = document.getElementById(this.containerId).querySelector('.hor-scroll');
+        if (this.wrapperScrollBar) {
+            let scrollbarTop = this.wrapperScrollBar.parentNode.querySelector('.hor-scroll-top');
+            if (!scrollbarTop){
+                this.createDoubleScroll(this.wrapperScrollBar);
+                this.scrollbarTop.addEventListener('scroll', this.handlerSyncWrapperScrollBar, false);
+                this.wrapperScrollBar.addEventListener('scroll', this.handlerSyncScrollBarTop, false);
+                const observer = new MutationObserver( this.handlerUpdateDoubleScrollWidth );
+                observer.observe(this.wrapperScrollBar, { childList: true, subtree: true });
+            }
+            this.updateDoubleScrollWidth();
+        }
+    }
+
+    // ensure varienGrid events propagation
+    getOldCallback(callbackName) {
+        return this.oldCallbacks[callbackName] ? this.oldCallbacks[callbackName] : Prototype.emptyFunction;
+    }
+
+    setOldCallback(callbackName, callback) {
+        this.oldCallbacks[callbackName] = callback;
+    }
+
+    onGridInit(grid) {
+        this.initDoubleScroll();
+        this.getOldCallback('init')(grid);
+    }
+
+    createDoubleScroll() {
+        let scrollbarTop = document.createElement('div');
+        scrollbarTop.classList.add('hor-scroll-top');
+        scrollbarTop.appendChild(document.createElement('div'));
+        scrollbarTop.style.overflow = 'auto';
+        scrollbarTop.style.overflowY = 'hidden';
+        scrollbarTop.firstChild.style.height = '0';
+        scrollbarTop.firstChild.style.paddingTop = '1px';
+        scrollbarTop.firstChild.appendChild(document.createTextNode('\xA0'));
+        this.wrapperScrollBar.parentNode.insertBefore(scrollbarTop, this.wrapperScrollBar);
+
+        this.scrollbarTop = scrollbarTop;
+    }
```

Comment:
I've read this file in more detail, the only thing that I'd try to do, in order to avoid as much javascript as possible, is to have this div directly into the `widget/grid.phtml` it should work ;-)

Reply:
I don't think it's so better only for save a few lines because if someone overwrites grid.phtml we lost it.

If you care about its size we can minify it as mage/adminhtml/grid-double-scroll.min.js and distribute also the original source mage/adminhtml/grid-double-scroll.js and can be done to others files like grid.js.

More javascript is not necessary bad if it's usefull, IMHO this is a usefull feature because if the columns don't fit in the window and I want to view the last column in a particular row, I have to scroll down to the end and then up to that row.
If we compare to other modern backends like Magento 2, our JS/Memory footprint is better

Reply:
I really woulnd't compare anything to M2 and I still believe, also from a performance standpoint, there's no need for element injection in js

Reply:
Okay, let's leave this decision pending for now

Reply:
also, grid.phtml is already modified by this PR

Reply:
have you tested if works correctly when there is more grids like in product view (related, upsell etc..) ?

Reply:
This script, as it is, is independent from the template. The only requirement is that the table is wrapped inside a div.hor-scroll, so it would also work on third-party implementations with a custom template that not include div.hor-scroll-top

Reply:
at the moment the script is called by a modified version of the template so it wouldn't work. I understand that it is callable by any situation but I'll never agree that many lines of javascript are better than 1 line of html + 1 line of css, both for readability and performance.
people may have a customized grid.html (doubt it but possible) but they could also have customized layout xml so the script won't be loaded, both approaches have pros and cons but the script is not 100% better.

Reply:
> at the moment the script is called by a modified version of the template so it wouldn't work

Then the script should be modified to work with this external div, with all handling of the events relative to the current grid

> people may have a customized grid.html (doubt it but possible) but they could also have customized layout xml so the script won't be loaded, both approaches have pros and cons but the script is not 100% better.

This is not necessarily true, the script should always be injected as grid.js unless explicitly removed in a layout update. 
IMHO I prefer to keep grid.phtml clean without a div that is not clear what it does, in terms of performance it doesn't change anything. 

But these are just details, the priority for me is that PR should be tested and reviewed on its implementation solution with interface contract, naming convention, etc.